src/ctapipe/io/datawriter.py

```diff
@@ -297,50 +289,180 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_value, exc_traceback):
         self.finish()
 
-    def __call__(self, event: ArrayEventContainer):
+    def __call__(self, event: SubarrayEventContainer):
         """
         Write a single event to the output file.
         """
         self._at_least_one_event = True
         self.log.debug("WRITING EVENT %s", event.index)
 
-        self._write_trigger(event)
-        # write fixed pointing only for simulation, observed data will have monitoring
-        if self._is_simulation:
-            self._write_constant_pointing(event)
+        self._write_simulation_subarray(event)
+        self._write_dl0_subarray(event)
+        self._write_dl2_subarray(event)
+
+        for tel_event in event.tel.values():
+            self._write_telescope_event(tel_event)
 
+    def _write_simulation_subarray(self, event):
         if event.simulation is not None and event.simulation.shower is not None:
             self._writer.write(
                 table_name="simulation/event/subarray/shower",
                 containers=[event.index, event.simulation.shower],
             )
 
-            for tel_id, sim in event.simulation.tel.items():
-                table_name = self.table_name(tel_id)
-                tel_index = _get_tel_index(event, tel_id)
-                self._writer.write(
-                    f"simulation/event/telescope/impact/{table_name}",
-                    [tel_index, sim.impact],
-                )
+    def _write_dl0_subarray(self, event):
+        self._writer.write(
+            table_name="dl0/event/subarray/trigger",
+            containers=[event.index, event.dl0.trigger],
+        )
 
-        if self.write_r1_waveforms:
-            self._write_r1_telescope_events(event)
+    def _write_telescope_event(self, tel_event):
+        table_name = self.table_name(tel_event.index.tel_id)
+        self._writer.write(
+            f"simulation/event/telescope/impact/{table_name}",
+            [tel_event.index, tel_event.simulation.impact],
+        )
+
+        if tel_event.simulation is not None:
```

Comment:
Why is `is not None` needed here?

Reply:
Because observed data do not have simulation information